### DISCUSSION TITLE
Containerlab on macOS - containerlab

### DISCUSSION THREAD

**Discussion by giscus** [srl-labs/containerlab#2333] [comments]
# Containerlab on macOS - containerlab

Containerlab deploys docker based meshed lab topologies

https://containerlab.dev/macos/

**Comment by addiekaul** [comment]
I'd like to provide information on Juniper cRPD which has been availble for Arm64 since Dec 2022, please find links Juniper cRPD Arm64 pre-requisites and also links for downloading cRPD, I have been using cRPD Arm version with containerlabs since couple of years without any issues, appreciate if you can update the wiki for macOS documentation.

https://www.juniper.net/documentation/us/en/software/crpd/crpd-deployment/topics/concept/crr-system-requirement-docker.html
https://www.juniper.net/documentation/us/en/software/junos/release-notes/21.4/junos-release-notes-21.4r1/topics/new-features/feature-descriptions/additional-features-7.html

https://support.juniper.net/support/downloads/?p=crpd
https://www.juniper.net/documentation/us/en/quick-start/software/crpd/crpd-quick-start/topics/concept/step-1-crpd.html

**Reply by hellt** [reply]
good call, I added a note about control-plane only cRPD option for arm64
it will be rendered with the next clab release

**Reply by PanWYao** [reply]
Just can't download the image..

**Comment by isenwulf** [comment]
I'm trying to get Docker outside of Docker working and I'm running into problems with the bind mounts. Specifically, `/run/docker/netns`, `/var/lib/docker`, etc. don't exist on my Mac. Using OrbStack to run Docker if that makes a difference.

**Reply by hellt** [reply]
How do you launch the container? 
And what orb machine os are you using? Did you install docker in it?

**Reply by isenwulf** [reply]
Here's my devcontainer.json:

```json
{
    "image": "ghcr.io/srl-labs/containerlab/devcontainer-dood-slim:latest", 
    "runArgs": [
        "--network=host",
        "--pid=host",
        "--privileged"
    ],
    "mounts": [
        "type=bind,src=/run/docker/netns,dst=/run/docker/netns",
        "type=bind,src=/var/lib/docker,dst=/var/lib/docker",
        "type=bind,src=/lib/modules,dst=/lib/modules"
    ],
    "workspaceFolder": "${localWorkspaceFolder}",
    "workspaceMount": "source=${localWorkspaceFolder},target=${localWorkspaceFolder},type=bind"
}
```
Trying to start the container from VSCode with `Dev Containers: Rebuild and Reopen in Container`.

> And what orb machine os are you using? Did you install docker in it?

I'm using the default VM that Orb creates when it first runs for handling Docker. Do I need to make a separate VM under Linux > Machines in Orb?

For reference here's the exact error message:

`docker: Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /run/docker/netns.`

**Reply by isenwulf** [reply]
Oh, it looks like you deleted the netns mount in your latest commits for your example labs. I deleted that line and I'm able to build a dev container instance. I guess we don't need that?

**Reply by hellt** [reply]
Correct. Instead of the latest tag use 0.61.0

**Reply by sjurtf** [reply]
I'm also having the same issue when using DOOD. I've tried versions 0.61.0, and also 0.60.0 but with the same outcome.

MacOS 14.7
Docker Desktop 4.37.2 (latest) 


```
Status 400: invalid mount config for type \"bind\": bind source path does not exist: /run/docker/netns
```

**Reply by UchihaItachiSama** [reply]
> I'm also having the same issue when using DOOD. I've tried versions 0.61.0, and also 0.60.0 but with the same outcome.
> 
> MacOS 14.7 Docker Desktop 4.37.2 (latest)
> 
> ```
> Status 400: invalid mount config for type \"bind\": bind source path does not exist: /run/docker/netns
> ```

I was seeing the issue on dood-slim 0.60.1, after updating the version to use 0.63.3 it seems to work without any issues

**Comment by manugeo4** [comment]
Can we get support for apple container instead of docker. More details about apple container below.
https://github.com/apple/container
https://developer.apple.com/videos/play/wwdc2025/346/

**Comment by nunomcruz** [comment]
I managed to run srlinux, arista and Mikrotik using vrnetlab on macOS 26.3 running on a M2 Max using Orbstack with a VM as per instructions. For Mikrotik I've used the VDI (arm64) image, converted it to vmdk using vbox-img tool from virtual box and using that image in vrnetlab. It worked surprisingly well!
